UNIQUE constraint failed: snapshot_excludes.hash, snapshot_excludes.path

Version: v0.2.9 / e5f58fd65ce4ec14aa493164ba56d59a1da9641a

```
$ RUST_BACKTRACE=1 cargo run -- -r "sftp:user@example:/path/to/repo" --password-file=/path/to/key
[…]
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 6.58s
     Running `target/debug/redu -r "sftp:user@example:/path/to/repo" --password-file=/path/to/key`
Logging to "/home/user/.local/share/redu/logs/2024-09-16T22:49:59.596074327+02:00.log"
 (●    ) Getting restic config                                                                                                 Using cache file "/home/user/.cache/redu/aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa.db"
 (●    ) Fetching repository snapshot list                                                                                     Error: error in db thread

Caused by:
    0: UNIQUE constraint failed: snapshot_excludes.hash, snapshot_excludes.path
    1: Error code 1555: A PRIMARY KEY constraint failed

Stack backtrace:
   0: anyhow::error::<impl core::convert::From<E> for anyhow::Error>::from
             at /home/user/.cargo/registry/src/index.crates.io-6f17d22bba15001f/anyhow-1.0.86/src/backtrace.rs:27:14
   1: core::ops::function::FnOnce::call_once
             at /rustc/04a318787b39732e306faf5ef6dc584990f4f417/library/core/src/ops/function.rs:250:5
   2: core::result::Result<T,E>::map_err
             at /rustc/04a318787b39732e306faf5ef6dc584990f4f417/library/core/src/result.rs:854:27
   3: redu::sync_snapshots::{{closure}}::{{closure}}
             at ./src/main.rs:247:17
   4: std::sys::backtrace::__rust_begin_short_backtrace
             at /rustc/04a318787b39732e306faf5ef6dc584990f4f417/library/std/src/sys/backtrace.rs:154:18
   5: std::thread::Builder::spawn_unchecked_::{{closure}}::{{closure}}
             at /rustc/04a318787b39732e306faf5ef6dc584990f4f417/library/std/src/thread/mod.rs:522:17
   6: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/04a318787b39732e306faf5ef6dc584990f4f417/library/core/src/panic/unwind_safe.rs:272:9
   7: std::panicking::try::do_call
             at /rustc/04a318787b39732e306faf5ef6dc584990f4f417/library/std/src/panicking.rs:557:40
   8: __rust_try
   9: std::panicking::try
             at /rustc/04a318787b39732e306faf5ef6dc584990f4f417/library/std/src/panicking.rs:520:19
  10: std::panic::catch_unwind
             at /rustc/04a318787b39732e306faf5ef6dc584990f4f417/library/std/src/panic.rs:348:14
  11: std::thread::Builder::spawn_unchecked_::{{closure}}
             at /rustc/04a318787b39732e306faf5ef6dc584990f4f417/library/std/src/thread/mod.rs:521:30
  12: core::ops::function::FnOnce::call_once{{vtable.shim}}
             at /rustc/04a318787b39732e306faf5ef6dc584990f4f417/library/core/src/ops/function.rs:250:5
  13: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at /rustc/04a318787b39732e306faf5ef6dc584990f4f417/library/alloc/src/boxed.rs:2453:9
  14: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at /rustc/04a318787b39732e306faf5ef6dc584990f4f417/library/alloc/src/boxed.rs:2453:9
  15: std::sys::pal::unix::thread::Thread::new::thread_start
             at /rustc/04a318787b39732e306faf5ef6dc584990f4f417/library/std/src/sys/pal/unix/thread.rs:105:17
  16: start_thread
  17: clone3
```

last lines from the log:
```
[…]
20:50:49 [INFO] redu::restic: running ""restic" "--repo" "sftp:user@example:/path/to/repo" "--password-file" "/path/to/key" "--json" "ls" "fdf9fdc755120d9ccaf87f3a47d305b88cc340b58bb9996a353cefce7b089f93"" (pid 29182)
20:50:50 [INFO] redu::restic: finished pid 28893
20:50:50 [INFO] redu: snapshot fetched in 25.057822208s (24fb842)
20:50:50 [INFO] redu::restic: running ""restic" "--repo" "sftp:user@example:/path/to/repo" "--password-file" "/path/to/key" "--json" "ls" "f7674983aeff1c65ab56f6d6656d12339e5fa7b5fbb30719fbd08dd41068265a"" (pid 29198)
20:50:50 [INFO] redu: waited 8.291456486s to save snapshot (891462 files)
20:50:50 [ERROR] sqlite: (1555) abort at 12 in [INSERT INTO snapshot_excludes (hash, path) VALUES (?, ?)]: UNIQUE constraint failed: snapshot_excludes.hash, snapshot_excludes.path
```

Cool project!

--- a/benches/cache.rs
+++ b/benches/cache.rs
@@ -25,7 +25,9 @@ pub fn criterion_benchmark(c: &mut Criterion) {
                 "/home/user".to_string(),
                 "/etc".to_string(),
                 "/var".to_string(),
-            ],
+            ]
+            .into_iter()
+            .collect(),
             hostname: Some("foo.com".to_string()),
             username: Some("user".to_string()),
             uid: Some(123),
@@ -34,8 +36,12 @@ pub fn criterion_benchmark(c: &mut Criterion) {
                 ".cache".to_string(),
                 "Cache".to_string(),
                 "/home/user/Downloads".to_string(),
-            ],
-            tags: vec!["foo_machine".to_string(), "rewrite".to_string()],
+            ]
+            .into_iter()
+            .collect(),
+            tags: vec!["foo_machine".to_string(), "rewrite".to_string()]
+                .into_iter()
+                .collect(),
             original_id: Some("fefwfwew".to_string()),
             program_version: Some("restic 0.16.0".to_string()),
         };
@@ -63,7 +69,9 @@ pub fn criterion_benchmark(c: &mut Criterion) {
                     "/home/user".to_string(),
                     "/etc".to_string(),
                     "/var".to_string(),
-                ],
+                ]
+                .into_iter()
+                .collect(),
                 hostname: Some("foo.com".to_string()),
                 username: Some("user".to_string()),
                 uid: Some(123),
@@ -72,8 +80,12 @@ pub fn criterion_benchmark(c: &mut Criterion) {
                     ".cache".to_string(),
                     "Cache".to_string(),
                     "/home/user/Downloads".to_string(),
-                ],
-                tags: vec!["foo_machine".to_string(), "rewrite".to_string()],
+                ]
+                .into_iter()
+                .collect(),
+                tags: vec!["foo_machine".to_string(), "rewrite".to_string()]
+                    .into_iter()
+                    .collect(),
                 original_id: Some("fefwfwew".to_string()),
                 program_version: Some("restic 0.16.0".to_string()),
             }

--- a/src/cache/tests.rs
+++ b/src/cache/tests.rs
@@ -1,5 +1,6 @@
 use std::{
-    cmp::Reverse, convert::Infallible, env, fs, iter, mem, path::PathBuf,
+    cmp::Reverse, collections::HashSet, convert::Infallible, env, fs, iter,
+    mem, path::PathBuf,
 };
 
 use camino::{Utf8Path, Utf8PathBuf};
@@ -323,21 +324,23 @@ fn cache_snapshots_entries() {
             assert_eq!(s0.original_id, s1.original_id);
             assert_eq!(s0.program_version, s1.program_version);
 
-            let mut s0_paths: Vec<String> = s0.paths.to_vec();
+            let mut s0_paths: Vec<String> = s0.paths.iter().cloned().collect();
             s0_paths.sort();
-            let mut s1_paths: Vec<String> = s1.paths.to_vec();
+            let mut s1_paths: Vec<String> = s1.paths.iter().cloned().collect();
             s1_paths.sort();
             assert_eq!(s0_paths, s1_paths);
 
-            let mut s0_excludes: Vec<String> = s0.excludes.to_vec();
+            let mut s0_excludes: Vec<String> =
+                s0.excludes.iter().cloned().collect();
             s0_excludes.sort();
-            let mut s1_excludes: Vec<String> = s1.excludes.to_vec();
+            let mut s1_excludes: Vec<String> =
+                s1.excludes.iter().cloned().collect();
             s1_excludes.sort();
             assert_eq!(s0_excludes, s1_excludes);
 
-            let mut s0_tags: Vec<String> = s0.tags.to_vec();
+            let mut s0_tags: Vec<String> = s0.tags.iter().cloned().collect();
             s0_tags.sort();
-            let mut s1_tags: Vec<String> = s1.tags.to_vec();
+            let mut s1_tags: Vec<String> = s1.tags.iter().cloned().collect();
             s1_tags.sort();
             assert_eq!(s0_tags, s1_tags);
         }
@@ -355,7 +358,9 @@ fn cache_snapshots_entries() {
             "/home/user".to_string(),
             "/etc".to_string(),
             "/var".to_string(),
-        ],
+        ]
+        .into_iter()
+        .collect(),
         hostname: Some("foo.com".to_string()),
         username: Some("user".to_string()),
         uid: Some(123),
@@ -364,8 +369,12 @@ fn cache_snapshots_entries() {
             ".cache".to_string(),
             "Cache".to_string(),
             "/home/user/Downloads".to_string(),
-        ],
-        tags: vec!["foo_machine".to_string(), "rewrite".to_string()],
+        ]
+        .into_iter()
+        .collect(),
+        tags: vec!["foo_machine".to_string(), "rewrite".to_string()]
+            .into_iter()
+            .collect(),
         original_id: Some("fefwfwew".to_string()),
         program_version: Some("restic 0.16.0".to_string()),
     };
@@ -375,7 +384,7 @@ fn cache_snapshots_entries() {
         time: mk_datetime(2025, 5, 12, 17, 00, 00),
         parent: Some("wat".to_string()),
         tree: "anothertree".to_string(),
-        paths: vec!["/home/user".to_string()],
+        paths: vec!["/home/user".to_string()].into_iter().collect(),
         hostname: Some("foo.com".to_string()),
         username: Some("user".to_string()),
         uid: Some(123),
@@ -384,8 +393,12 @@ fn cache_snapshots_entries() {
             ".cache".to_string(),
             "Cache".to_string(),
             "/home/user/Downloads".to_string(),
-        ],
-        tags: vec!["foo_machine".to_string(), "rewrite".to_string()],
+        ]
+        .into_iter()
+        .collect(),
+        tags: vec!["foo_machine".to_string(), "rewrite".to_string()]
+            .into_iter()
+            .collect(),
         original_id: Some("fefwfwew".to_string()),
         program_version: Some("restic 0.16.0".to_string()),
     };
@@ -395,13 +408,13 @@ fn cache_snapshots_entries() {
         time: mk_datetime(2023, 5, 12, 17, 00, 00),
         parent: None,
         tree: "fwefwfwwefwefwe".to_string(),
-        paths: vec![],
+        paths: HashSet::new(),
         hostname: None,
         username: None,
         uid: None,
         gid: None,
-        excludes: vec![],
-        tags: vec![],
+        excludes: HashSet::new(),
+        tags: HashSet::new(),
         original_id: None,
         program_version: None,
     };
@@ -460,13 +473,13 @@ fn lots_of_snapshots() {
             time: timestamp_to_datetime(i as i64).unwrap(),
             parent: None,
             tree: i.to_string(),
-            paths: vec![],
+            paths: HashSet::new(),
             hostname: None,
             username: None,
             uid: None,
             gid: None,
-            excludes: vec![],
-            tags: vec![],
+            excludes: HashSet::new(),
+            tags: HashSet::new(),
             original_id: None,
             program_version: None,
         };

--- a/src/restic.rs
+++ b/src/restic.rs
@@ -3,6 +3,7 @@ use core::str;
 use std::os::unix::process::CommandExt;
 use std::{
     borrow::Cow,
+    collections::HashSet,
     ffi::OsStr,
     fmt::{self, Display, Formatter},
     io::{self, BufRead, BufReader, Lines, Read},
@@ -313,7 +314,7 @@ pub struct Snapshot {
     #[serde(default)]
     pub parent: Option<String>,
     pub tree: String,
-    pub paths: Vec<String>,
+    pub paths: HashSet<String>,
     #[serde(default)]
     pub hostname: Option<String>,
     #[serde(default)]
@@ -323,9 +324,9 @@ pub struct Snapshot {
     #[serde(default)]
     pub gid: Option<u32>,
     #[serde(default)]
-    pub excludes: Vec<String>,
+    pub excludes: HashSet<String>,
     #[serde(default)]
-    pub tags: Vec<String>,
+    pub tags: HashSet<String>,
     #[serde(default)]
     pub original_id: Option<String>,
     #[serde(default)]

